pkg/install/quota.go

```diff
@@ -0,0 +1,83 @@
+package install
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+//checkQuota checks usage quotas vs. resources required by cluster before cluster creation
+func (i *Installer) checkQuota(ctx context.Context) error {
+	//prepare usage client
+	spp := &i.doc.OpenShiftCluster.Properties.ServicePrincipalProfile
+	r, err := azure.ParseResourceID(i.doc.OpenShiftCluster.ID)
+	if err != nil {
+		return err
+	}
+	conf := auth.NewClientCredentialsConfig(spp.ClientID, string(spp.ClientSecret), spp.TenantID)
+	authorizer, err := conf.Authorizer()
+	if err != nil {
+		return err
+	}
+	usageClient := compute.NewUsageClient(r.SubscriptionID, authorizer)
+	return i.calculateAndCompareQuota(ctx, usageClient)
+}
+
+func (i *Installer) calculateAndCompareQuota(ctx context.Context, uc compute.UsageClient) error {
+	//master nodes are already included in the declaration below
+	requiredResources := map[string]int{
+		"cores":              24,
+		"virtualMachines":    3,
+		"standardDSv3Family": 3,
+		"PremiumDiskCount":   3,
+	}
+	if api.VMSizeStandardD8sV3 != i.doc.OpenShiftCluster.Properties.MasterProfile.VMSize {
+		//if master node type changed error out
+		return fmt.Errorf("Unexpected Master node VMSize %s, expected %s", i.doc.OpenShiftCluster.Properties.MasterProfile.VMSize, api.VMSizeStandardD8sV3)
+	}
+	//worker node resource calculation
+	for _, w := range i.doc.OpenShiftCluster.Properties.WorkerProfiles {
+		i.log.Infof("WorkerProfile VMSize %s\n", w.VMSize)
+		requiredResources["virtualMachines"] = requiredResources["virtualMachines"] + w.Count
+		requiredResources["PremiumDiskCount"] = requiredResources["PremiumDiskCount"] + w.Count
+		switch w.VMSize {
+		case api.VMSizeStandardD2sV3:
+			requiredResources["standardDSv3Family"] = requiredResources["standardDSv3Family"] + w.Count
+			requiredResources["cores"] = requiredResources["cores"] + (w.Count * 2)
+		case api.VMSizeStandardD4sV3:
+			requiredResources["standardDSv3Family"] = requiredResources["standardDSv3Family"] + w.Count
+			requiredResources["cores"] = requiredResources["cores"] + (w.Count * 4)
+		case api.VMSizeStandardD8sV3:
+			requiredResources["standardDSv3Family"] = requiredResources["standardDSv3Family"] + w.Count
+			requiredResources["cores"] = requiredResources["cores"] + (w.Count * 8)
+		default:
+			//will only happen if pkg/api verification allows new VMSizes
+			return fmt.Errorf("Unexpected Worker node VMSize %s", w.VMSize)
+		}
+
+	}
+
+	usages, err := uc.List(ctx, i.doc.OpenShiftCluster.Location)
+	if err != nil {
+		return err
+	}
+	//check requirements vs. usage
+	for _, usage := range usages {
+		required, present := requiredResources[*usage.Name.Value]
```

Comment:
I think you need to iterate over `requiredResources` and see if all keys from that map are present in `usages`. And error, if usage misses a required key.

Maybe I'm wrong, but currently if `usages` doesn't have an item with, let's say, `*usage.Name.Value == "PremiumDiskCount"` the code will hapily continue execution.

Reply:
Yep, just tried to modify tests and this one passes (I think it should fail):

```go
{
	name: "not enough usage info to calculate",
	mocks: func(tt *test, uc *mock_compute.MockUsageClient) {
		uc.EXPECT().
			List(ctx, "ocLocation").
			Return([]compute.Usage{}, nil)
	},
},
```

Reply:
+1.  Sort the resource names before you iterate them so that the behaviour is stable (map iteration order is undefined in Golang).